File: web/packages/client/src/modbus-types.ts
/**
 * @file modbus_types.ts
 * @brief TypeScript definitions for Modbus-related enums, structs, and types
 *        extracted from C++ header files (ModbusTypes.h, ModbusManager.h,
 *        ModbusRTU.h, ModbusLogicEngine.h, Modbus.h).
 *
 * Naming Conventions:
 * - Enums: Capitalized, prefixed with E_
 * - Structs: Capitalized, prefixed based on context (Modbus related -> MB_)
 * - Types (aliases, interfaces for views): Capitalized, prefixed with T_
 *
 * Grouping:
 * - Common: Types/Enums used across RTU and TCP contexts.
 * - RTU: Types/Enums specific to Modbus RTU implementation.
 * - TCP: Types/Enums specific to Modbus TCP implementation (less direct structs found).
 * - Logic Engine: Types/Enums specific to the Modbus Logic Engine feature.
 */

// --- Common Constants (derived from C++ headers or configuration) ---
// Note: Actual values may differ based on the specific config-modbus.h
export const MAX_ADDRESSES_PER_SLAVE = 128; // Assuming typical config value for array sizes in SlaveData
export const MAX_PENDING_OPERATIONS = 64;   // Assuming typical config value for operation queue size
export const MAX_MODBUS_SLAVES = 16;        // Assuming typical config value for Manager/ModbusRTU slave data array size
export const MAX_MODBUS_COMPONENTS = 32;    // Assuming typical config value for ModbusTCP mapping storage
export const MAX_READ_BLOCKS = 10;          // Assuming typical config value for RTU_Base read block storage
export const MAX_REGISTERS = 128;           // Assuming typical config value for RTU_Base register array size


// Bitmask flags for MB_OPERATION (derived from C++ TEST/SET_BIT_TO usage)
// Note: Bit positions assumed based on typical usage, verify with C++ macros if available
export const OP_USED_BIT = 0;         // Flag indicating if the operation slot is used
export const OP_HIGH_PRIORITY_BIT = 1; // Flag indicating high priority
export const OP_IN_PROGRESS_BIT = 2;   // Flag indicating operation is currently being processed
export const OP_BROADCAST_BIT = 3;     // Flag indicating if it's a broadcast operation
export const OP_SYNCHRONIZED_BIT = 4;  // Flag indicating internal state is synchronized with the intended operation


// Bitmask flags for MB_VALUE_ENTRY (derived from C++ CBI/SBI usage)
export const VALUE_USED_BIT = 0;       // Flag indicating if the value entry slot is used
export const VALUE_SYNCHRONIZED_BIT = 1; // Flag indicating value has been synchronized with the slave


// Bitmask flags for MB_READ_BLOCK
export const BLOCK_USED_BIT = 0;       // Flag indicating if the read block slot is used


// --- Common Enums ---

/**
 * @brief Modbus Function Codes.
 */
export enum E_FN_CODE {
    FN_ANY_FUNCTION_CODE = 0x00, // Only valid for server to register function codes
    FN_READ_COIL = 0x01,
    FN_READ_DISCR_INPUT = 0x02,
    FN_READ_HOLD_REGISTER = 0x03,
    FN_READ_INPUT_REGISTER = 0x04,
    FN_WRITE_COIL = 0x05,
    FN_WRITE_HOLD_REGISTER = 0x06,
    FN_READ_EXCEPTION_SERIAL = 0x07,
    FN_DIAGNOSTICS_SERIAL = 0x08,
    FN_READ_COMM_CNT_SERIAL = 0x0B,
    FN_READ_COMM_LOG_SERIAL = 0x0C,
    FN_WRITE_MULT_COILS = 0x0F,
    FN_WRITE_MULT_REGISTERS = 0x10,
    FN_REPORT_SERVER_ID_SERIAL = 0x11,
    FN_READ_FILE_RECORD = 0x14,
    FN_WRITE_FILE_RECORD = 0x15,
    FN_MASK_WRITE_REGISTER = 0x16,
    FN_R_W_MULT_REGISTERS = 0x17,
    FN_READ_FIFO_QUEUE = 0x18,
    FN_ENCAPSULATED_INTERFACE = 0x2B,
    FN_USER_DEFINED_41 = 0x41,
    FN_USER_DEFINED_42 = 0x42,
    FN_USER_DEFINED_43 = 0x43,
    FN_USER_DEFINED_44 = 4, // Note: C++ has 0x44
    FN_USER_DEFINED_45 = 5, // Note: C++ has 0x45
    FN_USER_DEFINED_46 = 6, // Note: C++ has 0x46
    FN_USER_DEFINED_47 = 7, // Note: C++ has 0x47
    FN_USER_DEFINED_48 = 8, // Note: C++ has 0x48
    FN_USER_DEFINED_64 = 0x64,
    FN_USER_DEFINED_65 = 0x65,
    FN_USER_DEFINED_66 = 0x66,
    FN_USER_DEFINED_67 = 0x67,
    FN_USER_DEFINED_68 = 0x68,
    FN_USER_DEFINED_69 = 0x69,
    FN_USER_DEFINED_6A = 0x6A,
    FN_USER_DEFINED_6B = 0x6B,
    FN_USER_DEFINED_6C = 0x6C,
    FN_USER_DEFINED_6D = 0x6D,
    FN_USER_DEFINED_6E = 0x6E,
    FN_NONE = 0xFF,
}

/**
 * @brief Modbus operation types (a simplified view of function codes).
 */
export enum E_MB_OP_TYPE {
    MB_READ_COIL,
    MB_READ_REGISTER,
    MB_WRITE_COIL,
    MB_WRITE_REGISTER,
    MB_WRITE_MULTIPLE_REGISTERS
}

/**
 * @brief Modbus operation status.
 */
export enum E_MB_OP_STATUS {
    MB_PENDING,
    MB_SUCCESS,
    MB_FAILED,
    MB_RETRYING
}

/**
 * @brief Modbus operation error codes - incorporates internal errors and eModbus errors.
 */
export enum E_MB_ERROR {
    // Standard Modbus Exception Codes (as defined by eModbus)
    Success = 0x00, // Also used for internal success
    IllegalFunction = 0x01,
    IllegalDataAddress = 0x02,
    IllegalDataValue = 0x03,
    ServerDeviceFailure = 0x04,
    Acknowledge = 0x05,
    ServerDeviceBusy = 0x06,
    NegativeAcknowledge = 0x07,
    MemoryParityError = 0x08,
    GatewayPathUnavailable = 0x0A,
    GatewayTargetNoResp = 0x0B,

    // Internal Operation/Queue Errors (remapped to avoid conflicts)
    OpNotReady = 0x10,           // Renamed from NotReady
    OpQueueFull = 0x11,          // Renamed from QueueFull
    OpClientQueueFull = 0x12,    // Renamed from ClientQueueFull
    OpExecutionFailed = 0x13,    // Renamed from ExecutionFailed
    OpInvalidParameter = 0x14,   // Renamed from InvalidParameter
    OpRetrying = 0x15,           // Renamed from Retrying
    OpMaxRetriesExceeded = 0x16, // Renamed from MaxRetriesExceeded

    // eModbus Specific Communication Errors
    Timeout = 0xE0,
    InvalidServer = 0xE1,
    CrcError = 0xE2, // RTU only
    FcMismatch = 0xE3,
    ServerIdMismatch = 0xE4,
    PacketLengthError = 0xE5,
    ParameterCountError = 0xE6,
    ParameterLimitError = 0xE7,
    RequestQueueFull = 0xE8, // eModbus client queue
    IllegalIpOrPort = 0xE9,
    IpConnectionFailed = 0xEA,
    TcpHeadMismatch = 0xEB,
    EmptyMessage = 0xEC,
    AsciiFrameError = 0xED,
    AsciiCrcError = 0xEE,
    AsciiInvalidChar = 0xEF,
    BroadcastError = 0xF0,
    UndefinedError = 0xFF // Other communication error
}

/**
 * @brief Filter types for operation identification without RTTI.
 */
export enum E_FILTER_TYPE {
    FILTER_DUPLICATE,
    FILTER_RATE_LIMIT,
    FILTER_PRIORITY,
    FILTER_LIFECYCLE,
    FILTER_CUSTOM
}

// --- Common Structs / Interfaces ---

/**
 * @brief Structure to hold Modbus registration details for a component.
 */
export interface MB_REGISTERS {
    /** Starting address (-1/0xFFFF indicates invalid/not set) */
    startAddress: number; // ushort
    /** Number of consecutive addresses */
    count: number; // ushort
    /** Slave ID of the device */
    slaveId: number; // ushort (Note: RTU_Base uses uint8_t, reconcile based on context)
    /** Type of Modbus object */
    type: E_FN_CODE; // E_FN_CODE
    /** Read/Write access (value likely from an E_ModbusAccess enum not provided) */
    access: number; // E_ModbusAccess (assumed number type in TS)
    /** ID of the owning component (MAY NOT BE SET in data returned by mb_tcp_blocks) */
    componentId: number; // ushort
    /** Optional descriptive name for the register block */
    name: string | null; // const char* (string in TS)
    /** Optional group name for the register block */
    group: string | null; // const char* (string in TS)
    // ComponentFnPtr? Callback pointer is C++ specific, omit or describe conceptually
    // writeCallbackFn: any; // Conceptual callback function pointer
}

/**
 * @brief Structure to hold a Modbus operation. Representing data structure for the operation queue.
 * Optimized memory layout in C++ not directly applicable to TS interface.
 */
export interface MB_OPERATION {
    timestamp: number; // unsigned long (millis())
    token: number;     // uint32_t, Unique token for tracking response
    address: number;   // uint16_t, Modbus address
    value: number;     // uint16_t, Value for write operations / first value for reads
    quantity: number;  // uint16_t, Number of registers/coils for read/write multiple
    slaveId: number;   // uint8_t
    retries: number;   // uint8_t, Retry counter
    /**
     * @brief Flags byte (e.g., used, high priority, in progress, broadcast, synchronized).
     * Use OP_... constants to check bits.
     */
    flags: number;     // uint8_t
    type: E_FN_CODE;   // E_FN_CODE
    status: E_MB_OP_STATUS; // E_MB_OpStatus

    // C++ uses bitwise flags, define these as conceptual properties or helpers if needed
    // isUsed(): boolean; // Check flags using OP_USED_BIT
    // isHighPriority(): boolean; // Check flags using OP_HIGH_PRIORITY_BIT
    // isInProgress(): boolean; // Check flags using OP_IN_PROGRESS_BIT
    // isBroadcast(): boolean; // Check flags using OP_BROADCAST_BIT
    // isSynchronized(): boolean; // Check flags using OP_SYNCHRONIZED_BIT
}

/**
 * @brief Structure to represent a register or coil value entry stored in cache (e.g., in MB_SLAVE_DATA).
 */
export interface MB_VALUE_ENTRY {
    lastUpdate: number; // unsigned long (millis())
    address: number;    // uint16_t
    value: number;      // uint16_t (0 or 1 for coils, actual value for registers)
    /**
     * @brief Flags byte (e.g., used, synchronized).
     * Use VALUE_... constants to check bits.
     */
    flags: number;      // uint8_t

    // C++ uses bitwise flags, define these as conceptual properties or helpers if needed
    // isUsed(): boolean; // Check flags using VALUE_USED_BIT
    // isSynchronized(): boolean; // Check flags using VALUE_SYNCHRONIZED_BIT
}


// --- Common Types (Aliases) ---

/**
 * @brief A non-owning view of a collection of MB_REGISTERS blocks.
 */
export interface T_MODBUS_BLOCK_VIEW {
    /** Pointer to the first block. Never null if count > 0. */
    data: MB_REGISTERS[] | null; // Representing pointer to array
    /** Total number of blocks in the view. */
    count: number; // int
}

/**
 * @brief Callback type for when a Modbus RTU response is received.
 */
export type T_RESPONSE_CALLBACK = (slaveId: number) => void; // uint8_t slaveId

/**
 * @brief Callback type for notification when a register/coil value changes.
 */
export type T_ON_REGISTER_CHANGE_CALLBACK = (op: MB_OPERATION, oldValue: number, newValue: number) => void; // const ModbusOperation &op, uint16_t oldValue, uint16_t newValue

/**
 * @brief Callback type for notification when a write operation completes.
 */
export type T_ON_WRITE_CALLBACK = (op: MB_OPERATION) => void; // const ModbusOperation &op

/**
 * @brief Callback type for notification when a Modbus error occurs.
 */
export type T_ON_ERROR_CALLBACK = (op: MB_OPERATION, errorCode: number, errorMessage: string | null) => void; // const ModbusOperation &op, int errorCode, const char *errorMessage

/**
 * @brief Callback type for checking if operation already exists (used by filters).
 */
export type T_OPERATION_EXISTS_CALLBACK = (op: MB_OPERATION, context: any | null) => boolean; // const ModbusOperation &op, void *context


// --- RTU Specific Enums ---

/**
 * @brief Device State for RTU_Base class.
 */
export enum E_DEVICE_STATE {
    UNINITIALIZED,
    INITIALIZING,
    IDLE,
    RUNNING,
    ERROR
}

/**
 * @brief Initialization state for ModbusRTU client.
 */
export enum E_INIT_STATE {
    INIT_NOT_STARTED,
    INIT_SERIAL_STARTED,
    INIT_CLIENT_STARTED,
    INIT_READY,
    INIT_FAILED
}

// --- RTU Specific Structs / Interfaces ---

/**
 * @brief Struct for passing RTU update data via onMessage(void*).
 * Used for synchronous message passing - data copied immediately by receiver.
 */
export interface MB_RTU_UPDATE_DATA {
    slaveId: number;       // uint8_t, Original RTU Slave ID
    address: number;      // uint16_t, RTU Address OR Calculated TCP Address (depending on context)
    value: number;         // uint16_t
    functionCode: E_FN_CODE; // E_FN_CODE
}

/**
 * @brief Structure to represent a Modbus slave's data cache (coils and registers) managed by ModbusRTU.
 */
export interface MB_SLAVE_DATA {
    /** Array of coil entries (size MAX_ADDRESSES_PER_SLAVE) */
    coils: MB_VALUE_ENTRY[];
    /** Array of register entries (size MAX_ADDRESSES_PER_SLAVE) */
    registers: MB_VALUE_ENTRY[];
    /** Number of used coil entries */
    coilCount: number;     // uint8_t
    /** Number of used register entries */
    registerCount: number; // uint8_t
}

/**
 * @brief Structure for defining mandatory read blocks in Modbus RTU (used by RTU_Base).
 */
export interface MB_READ_BLOCK {
    /** Starting address of the block */
    startAddress: number;      // uint16_t
    /** Number of registers/coils in the block */
    count: number;             // uint16_t
    /** Modbus function code for reading (e.g., FN_READ_HOLD_REGISTER) */
    type: E_FN_CODE;             // E_FN_CODE
    /** Minimum interval between reads (ms) */
    readInterval: number; // unsigned long
    /** Timestamp of the last read attempt (millis()) */
    lastReadTime: number; // unsigned long
    /**
     * @brief Status flags (e.g., used).
     * Use BLOCK_USED_BIT constant.
     */
    flags: number;              // uint8_t

    // conceptual helpers
    // isUsed(): boolean; // Check flags using BLOCK_USED_BIT
}

/**
 * @brief Register state information. Represents the state and configuration of a single Modbus register/coil managed by RTU_Base.
 * Note: This was a C++ class, extracting data members as a type interface.
 */
export interface T_REGISTER_STATE {
    /** Type of register/coil (e.g., FN_READ_HOLD_REGISTER, FN_READ_COIL) */
    type: E_FN_CODE; // E_FN_CODE (Note: C++ uses RegisterState::E_RegType here, mapping assumed to E_FN_CODE or similar)
    /** Modbus address */
    address: number;    // uint16_t
    /** Current value */
    value: number;      // uint16_t
    /** Priority level for operations related to this register */
    priority: number;   // uint8_t

    // C++ methods like getBoolValue/setBoolValue/readFromDevice/writeToDevice are implementation details and omitted from interface.
    // getBoolValue(): boolean;
    // setBoolValue(boolValue: boolean): void;
}


// --- TCP Specific Types ---
// No unique structs/enums specific only to TCP were found other than the class itself,
// which primarily uses the common types like MB_REGISTERS and T_MODBUS_BLOCK_VIEW.

File: playground/lib/polymech-base/src/modbus/ModbusTypes.cpp
#include <ArduinoLog.h>
#include <modbus/ModbusTypes.h>
#include <modbus/ModbusRTU.h>
#include <components/RS485.h>

bool RTU_Base::triggerRTUWrite()
{
  RS485 *rs485 = (RS485 *)owner;
  if (rs485)
  {
    write(rs485->modbus);
    return true;
  }
  Log.errorln("Device %d: Failed to cast owner to RS485*! Cannot trigger RTU write.", slaveId);
  return false;
}

MB_Error RegisterState::readFromDevice(ModbusRTU &manager, uint8_t slaveId)
{
  switch (type)
  {
  case E_FN_CODE::FN_READ_INPUT_REGISTER:
  case E_FN_CODE::FN_READ_HOLD_REGISTER:
    return manager.readRegister(slaveId, address);
  case E_FN_CODE::FN_READ_COIL:
  case E_FN_CODE::FN_READ_DISCR_INPUT:
    return manager.readCoil(slaveId, address);
  }
  return MB_Error::UndefinedError;
}

MB_Error RegisterState::writeToDevice(ModbusRTU &manager, uint8_t slaveId, bool forceWrite)
{
  switch (type)
  {
  case E_FN_CODE::FN_WRITE_HOLD_REGISTER:
    return manager.writeRegister(slaveId, address, value, forceWrite);
  case E_FN_CODE::FN_WRITE_COIL:
    return manager.writeCoil(slaveId, address, getBoolValue());
  case E_FN_CODE::FN_READ_INPUT_REGISTER:
  case E_FN_CODE::FN_READ_DISCR_INPUT:
  case E_FN_CODE::FN_READ_HOLD_REGISTER:
  case E_FN_CODE::FN_READ_COIL:
    return MB_Error::IllegalDataAddress;
  default:
    Log.warningln("Attempted writeToDevice with unhandled/read-only type: %X", (int)type);
    return MB_Error::IllegalFunction;
  }
  return MB_Error::UndefinedError;
}

void RegisterState::printState(ModbusRTU &manager, uint8_t slaveId)
{
  switch (type)
  {
  case E_FN_CODE::FN_READ_INPUT_REGISTER:
    Log.noticeln("Input Register %d: %d (Priority: %d, %s)",
                 address, value, priority,
                 manager.isRegisterSynchronized(slaveId, address) ? "Synchronized" : "Not synchronized");
    break;

  case E_FN_CODE::FN_READ_HOLD_REGISTER:
    Log.noticeln("Holding Register %d: %d (Priority: %d, %s)",
                 address, value, priority,
                 manager.isRegisterSynchronized(slaveId, address) ? "Synchronized" : "Not synchronized");
    break;

  case E_FN_CODE::FN_READ_COIL:
    Log.noticeln("Coil %d: %s (Priority: %d, %s)",
                 address, getBoolValue() ? "ON" : "OFF", priority,
                 manager.isCoilSynchronized(slaveId, address) ? "Synchronized" : "Not synchronized");
    break;

  case E_FN_CODE::FN_READ_DISCR_INPUT:
    Log.noticeln("Discrete Input %d: %s (Priority: %d, %s)",
                 address, getBoolValue() ? "ON" : "OFF", priority,
                 manager.isCoilSynchronized(slaveId, address) ? "Synchronized" : "Not synchronized");
    break;

  default:
    Log.noticeln("Register Address %d: Value %d (Type: %X, Priority: %d, %s)",
                 address, value, (int)type, priority,
                 manager.isRegisterSynchronized(slaveId, address) ? "Synchronized" : "Not synchronized");
    break;
  }
}

DuplicateOperationFilter::DuplicateOperationFilter(ModbusRTU *rtu)
    : modbusRTU(rtu), operationTimeout(OPERATION_TIMEOUT)
{
  if (modbusRTU == nullptr)
  {
    Log.errorln("DuplicateOperationFilter created with null ModbusRTU pointer!");
  }
}

DuplicateOperationFilter::~DuplicateOperationFilter()
{
  // Nothing to clean up
}

bool DuplicateOperationFilter::filter(const ModbusOperation &op)
{
  // Check if this operation is already pending in the ModbusRTU queues
  // Use the stored modbusRTU pointer
  if (modbusRTU && modbusRTU->isOperationAlreadyPending(op))
  {
    // Log.traceln("Filter: Dropping duplicate operation (slave: %d, type: %d, address: %d)", op.slaveId, op.type, op.address);
    return false; // Filter out the operation
  }

  return true; // Let the operation through
}

RateLimitFilter::RateLimitFilter(unsigned long minIntervalMs)
    : minInterval(minIntervalMs), lastOperationTime(0) {}

bool RateLimitFilter::filter(const ModbusOperation &op)
{
  unsigned long now = millis();

  // Check if enough time has passed since the last operation
  if (now - lastOperationTime < minInterval)
  {
    Log.traceln("Filter: Rate limiting operation (slave: %d, type: %d, address: %d)",
                op.slaveId, op.type, op.address);
    return false; // Filter out the operation
  }

  // Update the last operation time
  lastOperationTime = now;
  return true; // Let the operation through
}

PriorityFilter::PriorityFilter() {}

bool PriorityFilter::filter(const ModbusOperation &op)
{
  return true;
}

bool PriorityFilter::adjustPriority(ModbusOperation &op)
{
  if (op.type == E_FN_CODE::FN_WRITE_COIL ||
      op.type == E_FN_CODE::FN_WRITE_HOLD_REGISTER ||
      op.type == E_FN_CODE::FN_WRITE_MULT_COILS ||
      op.type == E_FN_CODE::FN_WRITE_MULT_REGISTERS)
  {
    // Write operations are considered high priority
    // The actual priority value might be set elsewhere or this could return a priority level.
    // For now, returning true indicates it meets the high-priority condition.
    return true;
  }
  return false;
}

void RTU_Base::updateState(ModbusRTU &manager)
{
  if (state != UNINITIALIZED && lastResponseTime > 0 && millis() - lastResponseTime > responseTimeout)
  {
    if (state != ERROR)
    {
      errorCount++;
      setState(ERROR);
    }
    return; // If timeout occurred, don't process further state changes here
  }
  bool hasPendingOps = manager.hasPendingOperations(slaveId);
  switch (state)
  {
  case UNINITIALIZED:
    break;
  case INITIALIZING:
    if (!hasPendingOps)
    {
      setState(IDLE);
    }
    break;
  case IDLE:
    if (hasPendingOps)
    {
      setState(RUNNING);
    }
    break;
  case RUNNING:
    if (!hasPendingOps)
    {
      setState(IDLE);
    }
    break;
  case ERROR:
    break;
  }
}

bool RTU_Base::initialize(ModbusRTU &manager)
{
  if (state == UNINITIALIZED)
  {
    setState(INITIALIZING);
    write(manager);
    return true; // Return true indicating initialization was started
  }
  else
  {
    return false; // Return false as it wasn't in UNINITIALIZED state
  }
}

void RTU_Base::read(ModbusRTU &manager)
{
  if (state == RUNNING)
  {
    return;
  }
  unsigned long currentTime = millis();
  bool readQueued = false;
  int numBlocks = mandatoryReadBlocks.size();
  for (int i = 0; i < numBlocks; ++i)
  {
    ModbusReadBlock &block = mandatoryReadBlocks[i];

    if (!block.isUsed())
    {
      continue;
    }

    if (currentTime - block.lastReadTime >= block.readInterval)
    {
      MB_Error err = MB_Error::Success;
      switch (block.type)
      {
      case E_FN_CODE::FN_READ_HOLD_REGISTER:
        err = manager.readHoldingRegisters(slaveId, block.startAddress, block.count);
        break;
      case E_FN_CODE::FN_READ_INPUT_REGISTER:
        err = manager.readInputRegisters(slaveId, block.startAddress, block.count);
        break;
      case E_FN_CODE::FN_READ_COIL:
        err = manager.readCoils(slaveId, block.startAddress, block.count);
        break;
      case E_FN_CODE::FN_READ_DISCR_INPUT:
        err = manager.readDiscreteInputs(slaveId, block.startAddress, block.count);
        break;
      default:
        Log.warningln("Device %d: Mandatory read block has invalid type (%X). Skipping.", slaveId, (int)block.type);
        err = MB_Error::IllegalFunction; // Mark as error to prevent state change if this was the only block
        break;
      }

      if (err == MB_Error::Success)
      {
        readQueued = true;
        block.lastReadTime = currentTime;
      }
      else
      {
        if (err != MB_Error::OpNotReady)
        {
          Log.warningln("Device %d: Read queue is full. Skipping read for block - Start: %d, Count: %d.",
                        slaveId, block.startAddress, block.count);
        }
      }
    }
  }
  if (readQueued && state == IDLE)
  {
    setState(RUNNING);
  }
}

void RTU_Base::write(ModbusRTU &manager)
{
  if (state == RUNNING)
  {
    // Log.infoln("Device %d: Write operation already in progress. Skipping.", slaveId);
    // return;
  }
  bool writeQueued = false;
  for (int i = 0; i < registerCount; i++)
  {
    if (registers[i] != nullptr)
    {
      if (registers[i]->type == E_FN_CODE::FN_WRITE_HOLD_REGISTER ||
          registers[i]->type == E_FN_CODE::FN_WRITE_COIL)
      {
        MB_Error err = registers[i]->writeToDevice(manager, slaveId, registers[i]->priority == PRIORITY_HIGHEST);
        if (err == MB_Error::Success)
        {
          // Log.traceln("Device %d: Write operation queued for register %d (addr %d) | Value: %d.", slaveId, i, registers[i]->address, registers[i]->value);
          writeQueued = true;
        }
        else
        {
          Log.warningln("Device %d: Failed to queue write for register %d (addr %d). Error: %u",
                        slaveId, i, registers[i]->address, (unsigned int)err);
        }
      }
    }
    else
    {
      Log.errorln("Device %d: Register %d (addr %d) is not writable.", slaveId, i, registers[i]->address);
    }
  }
  if (writeQueued && state == IDLE)
  {
    setState(RUNNING);
  }
}

void RTU_Base::printState()
{
  for (int i = 0; i < registerCount; i++)
  {
    if (registers[i] != nullptr)
    {
      Log.traceln("Register %d: Type %d, Address %d, Value %d", i, registers[i]->type, registers[i]->address, registers[i]->value);
    }
  }
}

void RTU_Base::reset()
{
  Log.noticeln("Device %d: Resetting state...", slaveId);
  setState(UNINITIALIZED);
  lastResponseTime = 0;
  errorCount = 0;
  lastSyncTime = 0;
}

const char *RTU_Base::getStateString() const
{
  switch (state)
  {
  case UNINITIALIZED:
    return "Uninitialized";
  case INITIALIZING:
    return "Initializing";
  case IDLE:
    return "Idle";
  case RUNNING:
    return "Running";
  case ERROR:
    return "Error";
  default:
    return "Unknown";
  }
}

void RTU_Base::setState(E_DeviceState newState)
{
  if (state != newState)
  {
    E_DeviceState oldState = state;
    const char *oldStateStr = getStateString();
    state = newState;
    if ((oldState == ERROR) && (newState == IDLE || newState == RUNNING))
    {
      if (errorCount > 0)
      {
        errorCount = 0;
      }
    }
  }
}

void RTU_Base::handleResponseReceived()
{
  lastResponseTime = millis();
  if (state == ERROR)
  {
    setState(IDLE);
  }
  else if (state == IDLE || state == RUNNING || state == INITIALIZING)
  {
    // Log.traceln("Device %d received response in state %s.", slaveId, getStateString());
    // Log.traceln("Device %d received response in state %s. = %d", slaveId, getStateString());
    // printState();
  }
}

const RegisterState *RTU_Base::getRegisterByAddress(uint16_t address) const
{
  for (int i = 0; i < registerCount; i++)
  {
    if (registers[i] != nullptr && registers[i]->address == address)
    {
      return registers[i];
    }
  }
  return nullptr;
}

void RTU_Base::setOutputRegisterValue(uint16_t address, uint16_t value)
{
  bool found = false;
  for (int i = 0; i < registerCount; i++)
  {
    if (registers[i] != nullptr && registers[i]->address == address)
    {
      if (registers[i]->type == E_FN_CODE::FN_WRITE_HOLD_REGISTER || registers[i]->type == E_FN_CODE::FN_WRITE_COIL)
      {
        registers[i]->value = value;
        found = true;
        break;
      }
      else
      {
        Log.warningln("Device %d: Register with address %d found, but cannot set value - not a Holding/Coil type (%d).",
                      slaveId, address, registers[i]->type);
        found = true;
        break;
      }
    }
  }
  if (!found)
  {
    Log.warningln("Device %d: Writable register with address %d not found for setting value.", slaveId, address);
  }
  else
  {
    triggerRTUWrite();
  }
}

ModbusReadBlock *RTU_Base::addMandatoryReadBlock(uint16_t startAddress, uint16_t count, E_FN_CODE type, unsigned long interval)
{
  if (count == 0)
  {
    Log.warningln("Device %d: Cannot add mandatory read block with count 0.", slaveId);
    return nullptr;
  }
  if (type != E_FN_CODE::FN_READ_HOLD_REGISTER &&
      type != E_FN_CODE::FN_READ_INPUT_REGISTER &&
      type != E_FN_CODE::FN_READ_COIL &&
      type != E_FN_CODE::FN_READ_DISCR_INPUT)
  {
    Log.warningln("Device %d: Invalid type (%X) for mandatory read block.", slaveId, (int)type);
    return nullptr;
  }
  if (mandatoryReadBlocks.full())
  {
    Log.errorln("Device %d: Cannot add mandatory read block - maximum (%d) reached.", slaveId, MAX_READ_BLOCKS);
    return nullptr;
  }
  ModbusReadBlock newBlock(startAddress, count, type, interval);
  mandatoryReadBlocks.push_back(newBlock);
  Log.traceln("Device %d: Added mandatory read block - Start: %d, Count: %d, Type: %X, Interval: %lu ms",
              slaveId, startAddress, count, (int)type, interval);

  newBlock.setUsed(true);
  return &mandatoryReadBlocks.back();
}
void RTU_Base::onError(ushort errorCode, const char *errorMessage)
{
  lastErrorCode = errorCode;
}
void RTU_Base::onRegisterUpdate(uint16_t address, uint16_t newValue)
{
  bool updated = false;
  for (int i = 0; i < registerCount; ++i)
  {
    if (registers[i] != nullptr && registers[i]->address == address)
    {
      if (registers[i]->value != newValue)
      {
        registers[i]->value = newValue;
      }
      updated = true;
      break;
    }
  }
}
const char *modbusErrorToString(MB_Error error)
{
  switch (error)
  {
  // Standard Modbus Exceptions
  case MB_Error::Success:
    return "Success";
  case MB_Error::IllegalFunction:
    return "Illegal Function";
  case MB_Error::IllegalDataAddress:
    return "Illegal Data Address";
  case MB_Error::IllegalDataValue:
    return "Illegal Data Value";
  case MB_Error::ServerDeviceFailure:
    return "Server Device Failure";
  case MB_Error::Acknowledge:
    return "Acknowledge";
  case MB_Error::ServerDeviceBusy:
    return "Server Device Busy";
  case MB_Error::NegativeAcknowledge:
    return "Negative Acknowledge";
  case MB_Error::MemoryParityError:
    return "Memory Parity Error";
  case MB_Error::GatewayPathUnavailable:
    return "Gateway Path Unavailable";
  case MB_Error::GatewayTargetNoResp:
    return "Gateway Target Device Failed to Respond";

  // Internal Operation/Queue Errors
  case MB_Error::OpNotReady:
    return "Operation Not Ready";
  case MB_Error::OpQueueFull:
    return "ModbusRTU Operation Queue Full";
  case MB_Error::OpClientQueueFull:
    return "eModbus Client Queue Full";
  case MB_Error::OpExecutionFailed:
    return "Operation Execution Failed";
  case MB_Error::OpInvalidParameter:
    return "Invalid Parameter";
  case MB_Error::OpRetrying:
    return "Operation Retrying";
  case MB_Error::OpMaxRetriesExceeded:
    return "Max Retries Exceeded";

  // eModbus Specific Communication Errors
  case MB_Error::Timeout:
    return "Timeout";
  case MB_Error::InvalidServer:
    return "Invalid Server Response";
  case MB_Error::CrcError:
    return "CRC Error"; // RTU
  case MB_Error::FcMismatch:
    return "Function Code Mismatch";
  case MB_Error::ServerIdMismatch:
    return "Server ID Mismatch";
  case MB_Error::PacketLengthError:
    return "Packet Length Error";
  case MB_Error::ParameterCountError:
    return "Parameter Count Error";
  case MB_Error::ParameterLimitError:
    return "Parameter Limit Error";
  case MB_Error::RequestQueueFull:
    return "eModbus Request Queue Full"; // Same as OpClientQueueFull? Check enum vals
  case MB_Error::IllegalIpOrPort:
    return "Illegal IP or Port"; // TCP
  case MB_Error::IpConnectionFailed:
    return "IP Connection Failed"; // TCP
  case MB_Error::TcpHeadMismatch:
    return "TCP Header Mismatch"; // TCP
  case MB_Error::EmptyMessage:
    return "Empty Message Received";
  case MB_Error::AsciiFrameError:
    return "ASCII Frame Error"; // ASCII
  case MB_Error::AsciiCrcError:
    return "ASCII LRC Error"; // ASCII
  case MB_Error::AsciiInvalidChar:
    return "ASCII Invalid Character"; // ASCII
  case MB_Error::BroadcastError:
    return "Broadcast Error";

  case MB_Error::UndefinedError:
  default:
    return "Undefined or Unknown Error";
  }
}